# Moka Picard v1.2.1
 **The version of PicardTools used should be consistent with that used in GATK in the automated scripts**

## What does this app do?
The app runs dockerised versions of Picard (v2.22.8) and samtools (v1.13). Docker images have been saved as .tar.gz files which are loaded from 001_ToolsReferenceData.

The app runs modules from the Picard Tools suite to generate quality-control (QC) statistics from mapped/aligned reads. Specifically, this app (for capture panels):
* Calculates multiple summary statistic metrics for mapped reads (paired or unpaired) using Picard [CollectMultipleMetrics](https://broadinstitute.github.io/picard/command-line-overview.html#CollectMultipleMetrics).
* Calculates mappings metrics to determine the performance of the capture kit by assessing the coverage across all targets in the kit, using Picard [CalculateHsMetrics](https://broadinstitute.github.io/picard/command-line-overview.html#CollectHsMetrics).

And for amplicon panels:
* Calculates mappings metrics to determine the performance of the amplicon kit by assessing the coverage across all targets in the kit, using Picard [TargetedPcrMetrics](https://broadinstitute.github.io/picard/command-line-overview.html#TargetedPcrMetrics).

For more information on the Picard Tools suite see: 
* http://broadinstitute.github.io/picard/

## What are typical use cases for this app?
This app is designed to be run on aligned sequencing data, either as a standalone app or as part of a DNA Nexus workflow.

The QC metrics calculated by Picard tools and output by this app are informative of the quality of the sequence alignments produced by read mapping software (such as BWA).

The outputs of this app are to be displayed visually using [MultiQC](http://multiqc.info/), and assessed for inconsistencies accross the alignment summary. This summary contains the per-cycle base distribution, target enrichment and read duplication statistics.

## What data are required for this app to run?
The following files are required for this app to run:

**sorted_bam**:
A coordinate-sorted mapping file in BAM format (`*.bam`).

**fasta_index**:
A reference genome sequence index generated by "FASTA indexer (with Picard and Samtools)". This should be provided as a gzipped tar archive file (`*.fasta-index.tar.gz`).

**vendor_exome_bedfile**:
A BED file defining the regions designed to be captured. BED files can be found in 001_ToolsReferenceData:Data/BED. BED files for custom panels are created using the MokaBED app, using a list of gene symbols or accessions. For WES the capture BED file is used.

**Capture or amplicon boolean**
A boolean flag which denotes the run is a amplicon or capture based panel. 

Capture panels = True will run 
[CollectMultipleMetrics](https://broadinstitute.github.io/picard/command-line-overview.html#CollectMultipleMetrics) and [CalculateHsMetrics](https://broadinstitute.github.io/picard/command-line-overview.html#CollectHsMetrics). 

Capture panels = False will run [TargetedPcrMetrics](https://broadinstitute.github.io/picard/command-line-overview.html#TargetedPcrMetrics).

**remove_chr**
A boolean flag to denote if chr should be removed from the chromosome column in the intervals file. The intervals file is used by HsMetrics and collect_targeted_pcr_metrics. An example when this is set to false is the MokaSNP pipeline and TSO500 runs.

## What does this app output?
All Picard statistics files produced by this app are uploaded to a 'QC/' directory in the DNA Nexus project or working directory from which the app was called.

Picard CollectMultipleMetrics output files:
* `*.base_distribution_by_cycle*` - the base distribution per cycle
* `*.alignment_summary*` - a summary of the alignment
* `*.quality_by_cycle*` - the base quality per cycle
* `*.insert_size*` - metrics for validating library construction including the insert size distribution and read orientation of paired-end libraries
* `*.quality_distribution*` - the range of quality scores and the total numbers of bases corresponding to those scores

Picard CalculateHsMetrics output files:
* `*.hsmetrics.tsv` - general statistics about the enrichment process. 
* `*.pertarget_coverage.tsv` - the GC content and average coverage of each target in the kit.

Picard TargetedPcrMetrics output files:
* `*.targetPCRmetrics.txt` - A summary of the performance of the target amplicons
* `*.perTargetCov.txt` - A per-amplicon summary of %GC and coverage.


Detailed information about the metrics reported by all Picard suites can be found at the following page:
https://broadinstitute.github.io/picard/picard-metric-definitions.html

## How does this app work?
This app downloads the given input files and uses BAM and BED files to create a picard intervals_list file. Depending on the capture/amplicon flag either Picard TargetedPcrMetrics or both Picard CollectMultipleMetrics and Picard CalculateHsMetrics are then called. 

All output files are uploaded into the directory '/QC'.
